src/DataTransferObject.php

```diff
@@ -7,92 +7,206 @@
 use ReflectionClass;
 use ReflectionProperty;
 
-abstract class DataTransferObject
+/**
+ * Class DataTransferObject.
+ */
+abstract class DataTransferObject implements DtoContract
 {
     /** @var array */
     protected $exceptKeys = [];
 
     /** @var array */
     protected $onlyKeys = [];
 
+    /** @var Property[] | array */
+    protected $properties = [];
+
+    /** @var bool */
+    protected $immutable;
+
+    public static function mutable(array $parameters): self
+    {
+        return new static($parameters, false);
+    }
+
+    public static function immutable(array $parameters): self
```

Comment:
Returning `self` is a little problematic in PHP, because sub classes cannot return it. Here's an example:

```php
class A
{
    public function foo(): self {}
}

class B extends A
{
    public function foo(): self {}
}
```

This gives an error because the return type of `B::foo` should be `A`, instead it's `B`. That's why I always return the concrete class, to make it very clear what should be returned.

```php
class A
{
    public function foo(): A {}
}

class B extends A
{
    public function foo(): A {}
}
```

As soon as PHP 7.4 drops with better type variance, there are more possibilities, but for now this will have to do.

Reply:
"As soon as PHP 7.4 drops with better type variance," I've been waiting for that for years (that along with typed properties & generics).

What about this implementation:
```
    /**
     * @return static
     */
    public function mutable()
    {
        $this->immutable = false;
        return $this;
    }

    /**
     * @return static
     */
    public function immutable()
    {
        $this->immutable = true;
        return $this;
    }
```

Reply:
it will break the current api (that contains the immutable static method) though and set mutable by default

Reply:
i readded the static immutable method and marked it as deprecated